Monitored trip

br648 opened the pull request:
### Checklist

- [x] Appropriate branch selected _(all PRs must first be merged to `dev` before they can be merged to `master`)_
- [x] Any modified or new methods or classes have helpful JavaDoc and code is thoroughly commented
- [ ] The description lists all applicable issues this PR seeks to resolve
- [ ] The description lists any configuration setting(s) that differ from the default settings
- [x] All tests and CI builds passing

### Description

Responsible for creating/updating/deleting monitored trips. Create, update and delete pre hooks validate a user prior to carrying out related action. Create has an additional check on the number of monitored trips previously created. This is set to 5 and will prevent new monitored trips being created if this is to be exceeded.

Commit:
Initial monitor trip work

Commit:
Updated obj to string method

Commit:
Fix merge conflict

Commit:
Added maximum monitored trips check

Commit:
Merge branch 'trip-history' into monitored-trip

Commit:
Added user auth

Commit:
Fixed merge conflicts

Commit:
Comment updates

Commit:
Update get many to return entities based on user permissions

Commit:
Updated comments

Commit:
Removed redundant imports

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/auth/Auth0Connection.java:
```diff
@@ -80,6 +80,11 @@ public static void checkUserIsAdmin(Request req, Response res) {
     }
 
+    /** Check if the incoming user is an admin user */
+    public static boolean isUserAdmin(Auth0UserProfile user) {
```
Make method `checkUserIsAdmin` call `isUserAdmin`.

landonreed replied:
Actually, I wouldn't worry about making this change. I've added @br648's `isUserAdmin` method to #14 and am using the method in `checkUserIsAdmin` in that downstream PR. So I'd rather if we just leave that unchanged to avoid potential (even if unlikely) conflicts.

Commit:
Updated auth to use OTP user already in request

Commit:
Moved generic user and auth checks to Auth0 utils

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,79 @@
+import java.util.Set;
+
+/**
+ * A monitored trip represents
```
Finish this comment...

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,79 @@
+public class MonitoredTrip extends Model {
+
+    /**
+     * User Id. Id of user monitoring trip.
```
Say rather, "Mongo Id of user monitoring trip"

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,79 @@
+    public Set<String> days;
+
+    /**
+     * Specify if the monitored trip should be checked on a federal holiday.
```
US federal holiday

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,79 @@
+    /**
+     * The trips itinerary
+     */
+    public List<Itinerary> itinerary;
```
What was the decision on this @landonreed? Did we decide to focus on one itinerary or multiple?

binh-dam-ibigroup replied:
I think it is ok for now because from the UI, we query the itinerary without real-time updates, and `queryParams` contains the index of the active itinerary.

evansiroky replied:
This should be a single itinerary that is stored. When we check for differences we want to be checking against a single itinerary.

br648 replied:
Updated.

binh-dam-ibigroup reviewed (changes requested):
Please merge with the latest changes from branch `user-sub-types` or `basic-user-prefs`. There is some code introduced in this PR that is no longer needed when merging with those other branches.
There are some other minor JS doc comment changes.

Commit:
Fixed merge conflicts and addressed PR issues

Commit:
Corrected missed merge conflict

Commit:
Added special case to get many to manage OTP users

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -148,10 +154,23 @@ protected void buildEndPoint(ApiEndpoint baseEndPoint) {
+        if (isUserAdmin(requestingUser)) {
+            return persistence.getAll();
+        } else if (persistence.clazz == OtpUser.class) { 
+            Bson filter = Filters.and(eq("_id", requestingUser.otpUser.id));
```
Suggestion: You don't need to use `Filters.and` if there is just one condition. See https://github.com/ibi-group/otp-middleware/blob/4b0cb1ed711fb7ec834563340d2eafb61abba4d0/src/main/java/org/opentripplanner/middleware/auth/Auth0UserProfile.java#L44

br648 replied:
Updated accordingly

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -148,10 +154,23 @@ protected void buildEndPoint(ApiEndpoint baseEndPoint) {
+            return persistence.getFiltered(filter);
+        } else {
+            // FIXME assumes all non admin user collections will have a user id field
+            Bson filter = Filters.and(eq("userId", requestingUser.otpUser.id));
```
Same as above.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitorTripController.java:
```diff
@@ -0,0 +1,61 @@
+ * with Auth0 services using the hooks provided by {@link ApiController}.
+ */
+public class MonitorTripController extends ApiController<MonitoredTrip> {
+    private static final long MAXIMUM_ALLOWED_MONITORED_TRIPS = 5;
```
I don't mind having this as a constant for now. I defer to the other reviewers.

evansiroky replied:
Let's make this be obtained from the application config.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/TripHistoryController.java:
```diff
@@ -46,20 +46,15 @@
+
         OtpUser user = Persistence.otpUsers.getById(userId);
         if (user == null) {
             logMessageAndHalt(request, HttpStatus.FORBIDDEN_403, "Unknown user.");
```
Should this be 404?

br648 replied:
I think it should be 403, however, after reviewing I have removed this check as it replicates the check carried out by isValidUser.

binh-dam-ibigroup replied:
^ ^ I'll add this as a change to #17.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,79 @@
+public class MonitoredTrip extends Model {
+
+    /**
+     * User Id. Mongo Id of user monitoring trip.
```
Suggestion: Remove first sentence in comment.

br648 replied:
Updated.

binh-dam-ibigroup commented on src/test/java/org/opentripplanner/middleware/persistence/PersistenceUtil.java:
```diff
@@ -77,4 +80,63 @@ public static void deleteTripRequests(List<TripRequest> tripRequests) {
 
+    public static MonitoredTrip createMonitoredTrip(String userId) {
+        Set<String> days = new HashSet<>();
+        days.add("Monday");
```
Should the days lowercase or uppercase? See https://github.com/ibi-group/otp-middleware/blob/e85febf23a8accad01a28f49d26bcb63cc8053ad/src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java#L42

br648 replied:
I've changed to lowercase. I've also added a comment to the parameter in MonitoredTrip regarding case! Shame enums can't be used.

binh-dam-ibigroup reviewed (changes requested):
Minor refactors, then it's good to go for me.

Commit:
Added auth check to GetOne entity. Addressed PR issues

Commit:
Fixed merge conflict

Commit:
Corrected HTTP status

binh-dam-ibigroup reviewed (approved):
Nice job @br648 !

Commit:
Additional work around auth in particular to allow admin greater permissions

br648 commented:
@binh-dam-ibigroup I've had to do further auth changes as the restrictions were too great. Admins didn't have the required permissions.

Commit:
Fixed merge conflicts

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/auth/Auth0Connection.java:
```diff
@@ -155,21 +155,32 @@ public static boolean authDisabled() {
+        }
+
+        if (userId == null ||
+            (profile.otpUser != null && !profile.otpUser.id.equalsIgnoreCase(userId)) ||
```
Any reason for ignoring case on user Mongo ID? (Comment code accordingly)

br648 replied:
I have reduced to just "equals".

binh-dam-ibigroup reviewed (changes requested):
Please resolve conflicts and justify ignoring case when checking user Mongo ID.

Commit:
Removed ignore case when comparing user id

Commit:
Fixed merge conflict

binh-dam-ibigroup committed:
fix(JsonUtils): Fix error parsing Dates in OTP Itinerary objects and their children.

Commit:
Merge branch 'trip-history' into monitored-trip

Commit:
Merge branch 'trip-history' into monitored-trip

Commit:
Fixed failing tests

Commit:
Fixed merge conflicts

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -140,18 +145,36 @@ public void bind(final SparkSwagger restApi) {
+            Bson filter = Filters.eq("_id", requestingUser.otpUser.id);
+            return persistence.getFiltered(filter);
+        } else {
+            // FIXME assumes all non admin user collections will have a user id field
```
This seems like a reasonable design choice to me so I'd be fine with removing the FIXME.

br648 replied:
Updated.

evansiroky commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitorTripController.java:
```diff
@@ -0,0 +1,57 @@
+    /**
+     * Confirm that the maximum number of saved monitored trips has not been reached
+     */
+    private void reachedMaximum(String userId, Request request) {
```
Can we rename this to `verifyBelowMaxNumTrips`?

br648 replied:
Updated.

evansiroky commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,111 @@
+    notifyRouteChange (true/false) - Instead of notificationThresholds
+    arrivalDelayMinutesThreshold (int minutes) - Instead of notificationThresholds
+    departureDelayMinutesThreshold (int minutes) - Instead of notificationThresholds
+    notifyDelayToTrip (true/false)
```
For now, I think it's best to keep these to a minimum and specific fields similar to these recommended as class members.

br648 replied:
@landonreed @binh-dam-ibigroup can we decide what is needed so I can update please?

binh-dam-ibigroup replied:
We have not decided yet. I think you can leave as is because there is the TODO reminder for that.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/JsonUtils.java:
```diff
@@ -16,15 +15,22 @@
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+        }
+        return "{}";
```
If an error happens, shouldn't we return an error in JSON?

br648 replied:
Updated, but not sure on syntax. Will need reviewing.

evansiroky reviewed (changes requested):
Just a few small comments. I'll approve once we figure out how to store notification trigger settings.

evansiroky commented:
Adding the blocked tag as this PR depends on #17 and #15.

Commit:
Merge branch 'trip-history' into monitored-trip

Commit:
Moved max monitored trips to config file. Created method to provide int config param. Updated comments. Updated toJson to provide error details on failure. Reduced monitored trip so single itinerary.

Commit:
Fixed merge conflict

Commit:
Moved monitored trip tests into separate class

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/utils/JsonUtils.java:
```diff
@@ -16,15 +15,26 @@
 public class JsonUtils {
```
`HttpResponse` import is unused.

br648 replied:
Updated.

binh-dam-ibigroup commented on configurations/default/env.yml.tmp:
```diff
@@ -10,4 +10,5 @@ MONGO_DB_NAME: otp_middleware
-OTP_SERVER_PLAN_END_POINT: /otp/routers/default/plan
\ No newline at end of file
+OTP_SERVER_PLAN_END_POINT: /otp/routers/default/plan
+MAXIMUM_PERMITTED_MONITORED_TRIPS: 5
```
Is this parameter required? In `Main.java`, there is something that assigns a default value, however that is not taking effect when `MAXIMUM_PERMITTED_MONITORED_TRIPS` is not set in the configuration.

br648 replied:
@binh-dam-ibigroup yes this parameter is required. I have updated the method to use the default value if the parameter is missing from the configuration.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/spark/Main.java:
```diff
@@ -189,4 +187,22 @@ public static String getConfigPropertyAsText(String name, String defaultValue) {
+
+        try {
+            JsonNode node = getConfigProperty(name);
+            value = Integer.parseInt(node.asText());
```
If `MAXIMUM_PERMITTED_MONITORED_TRIPS` is not defined in the configuration, I think this should simply use the default value assigned above.

binh-dam-ibigroup replied:
If `MAXIMUM_PERMITTED_MONITORED_TRIPS` is not defined in the configuration, a `NullPointerException` is triggered that should be caught.

br648 replied:
Updated. Good spot.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/JsonUtils.java:
```diff
@@ -25,12 +25,16 @@
         }
-        return "{}";
+
+        return String.format("{Error: Unable to serialize object: %s. %s.", object.toString(), message);
```
This is invalid JSON. Ideally, we should just make a call to `JsonUtils#formatJSON`. See https://github.com/ibi-group/otp-middleware/blob/a323b679905673d781a19358acf679baad4ac3dd/src/main/java/org/opentripplanner/middleware/utils/JsonUtils.java#L98-L100

evansiroky reviewed (changes requested):
See comment about JSON.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,111 @@
+import org.opentripplanner.middleware.auth.Permission;
+import org.opentripplanner.middleware.otp.response.Itinerary;
+
+import java.util.List;
```
Remove unused import.

br648 replied:
Updated.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitorTripController.java:
```diff
@@ -0,0 +1,60 @@
+ * Implementation of the {@link ApiController} abstract class for managing monitored trips. This controller connects
+ * with Auth0 services using the hooks provided by {@link ApiController}.
+ */
+public class MonitorTripController extends ApiController<MonitoredTrip> {
```
Rename to `MonitoredTrip`

br648 replied:
Updated.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitorTripController.java:
```diff
@@ -0,0 +1,60 @@
+        = getConfigPropertyAsInt("MAXIMUM_PERMITTED_MONITORED_TRIPS", 5);
+
+    public MonitorTripController(String apiPrefix) {
+        super(apiPrefix, Persistence.monitoredTrip, "secure/monitortrip");
```
Rename to `secure/monitoredtrip`

br648 replied:
Updated.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/TripHistoryController.java:
```diff
@@ -135,4 +135,7 @@ private static Date getDate(Request request, String paramName, String paramValue
                 .atZone(ZoneId.systemDefault())
                 .toInstant());
     }
+
```
Remove these spaces?

br648 replied:
Updated.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitorTripController.java:
```diff
@@ -0,0 +1,60 @@
+import static com.mongodb.client.model.Filters.eq;
+import static org.opentripplanner.middleware.auth.Auth0Connection.isAuthorized;
+import static org.opentripplanner.middleware.spark.Main.getConfigPropertyAsInt;
+import static org.opentripplanner.middleware.spark.Main.getConfigPropertyAsText;
```
This import appears unused.

br648 replied:
Removed.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java:
```diff
@@ -1,6 +1,9 @@
 package org.opentripplanner.middleware.otp.response;
 
+import com.fasterxml.jackson.annotation.JsonIgnore;
```
Unused import.

br648 replied:
Removed.

binh-dam-ibigroup commented on src/main/java/org/opentripplanner/middleware/otp/response/Itinerary.java:
```diff
@@ -1,6 +1,9 @@
+import com.fasterxml.jackson.annotation.JsonIgnore;
 import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
+
+import java.util.ArrayList;
```
Unused import.

br648 replied:
Removed.

binh-dam-ibigroup reviewed (changes requested):
Please advise on whether the trip limit config parameter should be required or optional. Rename a controller, and remove some unused imports.

br648 commented:
> 
> 
> Please advise on whether the trip limit config parameter should be required or optional. Rename a controller, and remove some unused imports.

@binh-dam-ibigroup MAXIMUM_PERMITTED_MONITORED_TRIPS is required. I have updated the code to take this into account. I've removed all unused imports from the project.

Commit:
Removed all unused imports. Renamed controller. Updated getConfigPropertyAsInt to account for missing parameter.

Commit:
Refactored toJson to provide correct json response if an exception occurs

binh-dam-ibigroup reviewed (dismissed):
Good work @br648 !

Commit:
Fixed merge conflicts

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -140,18 +145,35 @@ public void bind(final SparkSwagger restApi) {
+        } else {
+            Bson filter = Filters.eq("userId", requestingUser.otpUser.id);
+            return persistence.getFiltered(filter);
+        }
```
I would like to see some comments accompanying this block of conditionals describing the intent. It may be obvious to all of us at this moment, but future readers may not fully understand the context for the logic we've embedded here. E.g., "If user is an admin, the context is presumed to be the admin dashboard, so we deliver all entities (e.g., monitored trips) for management or review" etc...

br648 replied:
updated

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -140,18 +145,35 @@ public void bind(final SparkSwagger restApi) {
         String id = getIdFromRequest(req);
-        return getObjectForId(req, id);
+        T object = getObjectForId(req, id);
+        if (!object.canBeManagedBy(requestingUser)) {
```
I think the name `canBeManagedBy` implies that the requesting user can update or delete the entity. I'm not sure if it applies to viewing the entity. Do we need a separate method for `getOne`?

br648 replied:
@landonreed I think you are right. I'm not sure how a separate method that would tie-in with the controller 'get one entity' endpoint. Perhaps the way forward is along the same lines as getMany?

landonreed replied:
Yea, I think you're right. If the object can be managed by the user, the user is obviously authorized. Else, the requesting userId must match the userId attached to the requested object.

br648 replied:
@landonreed - Thanks. Will update...

br648 replied:
@landonreed - After testing, it is the responsibility of the overriding class to define access to it's object. I have updated the method comment to clarify this. Therefore, I don't think the method should be changed.

landonreed commented on src/main/java/org/opentripplanner/middleware/persistence/Persistence.java:
```diff
@@ -74,7 +75,8 @@ public static void initialize () {
         tripRequests = new TypedPersistence(mongoDatabase, TripRequest.class);
         tripSummaries = new TypedPersistence(mongoDatabase, TripSummary.class);
-        // TODO Add other models...
+        monitoredTrip = new TypedPersistence(mongoDatabase, MonitoredTrip.class);
```
Please pluralize to match other collections.

br648 replied:
Updated.

landonreed commented on src/main/java/org/opentripplanner/middleware/spark/Main.java:
```diff
@@ -7,11 +7,8 @@
-import org.opentripplanner.middleware.controllers.api.AdminUserController;
-import org.opentripplanner.middleware.controllers.api.ApiUserController;
-import org.opentripplanner.middleware.controllers.api.TripHistoryController;
+import org.opentripplanner.middleware.controllers.api.*;
```
I think there are a few places where wildcard imports have been substituted for single imports. Could you change this back? This setting needs to be modified in your IDE settings (I think you can change the threshold for switching to wildcard to 9999 or something like that).

br648 replied:
Updated IDE and removed wildcard imports

landonreed commented on src/test/java/org/opentripplanner/middleware/persistence/PersistenceUtil.java:
```diff
@@ -83,6 +86,63 @@ public static void deleteTripRequests(List<TripRequest> tripRequests) {
+        monitoredTrip.leadTimeInMinutes = 30;
+        monitoredTrip.days = days;
+        monitoredTrip.excludeFederalHolidays = true;
+        monitoredTrip.queryParams = "userId=b46266f7-a461-421b-8e92-01d99b945ab0&fromPlace=28.54894%2C%20-81.38971%3A%3A28.548944048426772%2C-81.38970606029034&toPlace=28.53989%2C%20-81.37728%3A%3A28.539893820446867%2C-81.37727737426759&date=2020-05-05&time=12%3A04&arriveBy=false&mode=WALK%2CBUS%2CRAIL&showIntermediateStops=true&maxWalkDistance=1207&optimize=QUICK&walkSpeed=1.34&ignoreRealtimeUpd ...
```
userId should no longer be in query params since we switched to using the Auth token, right?

br648 replied:
Updated.

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/MonitoredTripController.java:
```diff
@@ -0,0 +1,59 @@
+
+    @Override
+    boolean preDeleteHook(MonitoredTrip monitoredTrip, Request req) {
+        // Authorization checks are done prior to this hook
```
It seems a little off that we would handle authorization differently for delete vs. update/create. Also, I noticed a potential bug in the isAuthorized check. I think it's possible that Auth0UserProfile could have both an apiUser and an otpUser. The way the check works now, I think it will throw a halt if (e.g.) the otpUser.id matches the monitored trip userId, but the apiUser.id does not.

br648 replied:
We would need to decide if the delete auth is to match the update/create auth... as all controllers will need to be updated to account for this.

I have separated the otpUser and apiUser userid checks.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,110 @@
+     *
+     * All use of this parameter will require equalsIgnoreCase() to avoid missing days.
+     */
+    public Set<String> days;
```
I think it might be more appropriate to have boolean fields for each day. @evansiroky?

br648 replied:
I've updated to boolean fields for each day.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,110 @@
+        boolean isManagingSelf = false;
+
+        if (user.otpUser != null) {
+            isManagingSelf = userId.equalsIgnoreCase(user.otpUser.id);
```
Not sure we should be using ignore case.

br648 replied:
Updated.

landonreed commented on src/main/java/org/opentripplanner/middleware/models/MonitoredTrip.java:
```diff
@@ -0,0 +1,110 @@
+    @Override
+    public boolean canBeManagedBy(Auth0UserProfile user) {
+        // If the user is attempting to update someone else's monitored trip, they must be admin.
+        boolean isManagingSelf = false;
```
Variable name feels awkward. What about: `belongsToUser`?

br648 replied:
Updated.

landonreed reviewed (changes requested):
A few changes for you, but otherwise looking good!

Commit:
Updated comments. Corrected import wildcards. Auth updates. Changed monitored trip day check.

br648 commented:
@landonreed A couple of points which I need feedback on before completing these review.

Commit:
Fixed merge conflicts with base branch

binh-dam-ibigroup reviewed (dismissed):
LGTM

Commit:
Updated comment for getOne method

landonreed commented on src/main/java/org/opentripplanner/middleware/controllers/api/ApiController.java:
```diff
@@ -162,26 +161,38 @@ protected void buildEndpoint(ApiEndpoint baseEndpoint) {
+            // If the required entity is of type 'OtpUser' the assumption is that a call is being made via the
+            // OtpUserController. Therefore, the request should be limited to return just the entity matching the
+            // requesting user.
             Bson filter = Filters.eq("_id", requestingUser.otpUser.id);
```
The filter should be on the field `userId`, right? Perhaps this should be factored into a method (it's used below and there may be other occurrences as well.

br648 replied:
@landonreed  - "_id" is correct. The filter works on the columns used in Mongo. For an OTP user, the userId is held under the _id column. I have added a method to return the objects based on the these filter requirements.

landonreed replied:
Right, ok. I think my confusion was around the fact that we're requesting "many" but filtering on the "_id" field, which by design should only return one result.

landonreed reviewed (dismissed):
Changes look good! I think it just needs one more tweak before merging (see [comment](https://github.com/ibi-group/otp-middleware/pull/21/files/4d29ba4735acc1847c790df5e145e4c8917fe4eb..76baee3085dda2f1d971897a18353c25bf8e068e#r450409161))

Commit:
Created method to return objects filtered by field name and value

br648 commented:
@binh-dam-ibigroup My latest commit dismissed your previous review and I think approval?

binh-dam-ibigroup commented:
> @binh-dam-ibigroup My latest commit dismissed your previous review and I think approval?

@br648 Still stands after your last commit.

evansiroky commented on src/main/java/org/opentripplanner/middleware/utils/JsonUtils.java:
```diff
@@ -4,27 +4,37 @@
-    private static Gson gson = new Gson();
+
+    // Gson is not suitable for parsing OTP itineraries (errors on parsing Dates).
+    // private static Gson gson = new Gson();
```
Let's remove this comment then.

br648 replied:
Updated.

evansiroky commented on src/test/java/org/opentripplanner/middleware/persistence/MonitoredTripPersistenceTest.java:
```diff
@@ -0,0 +1,45 @@
+ * number of {@link TypedPersistence} methods are tested here, but the HTTP endpoints defined in
+ * {@link org.opentripplanner.middleware.controllers.api.ApiController} are not themselves tested here.
+ */
+public class MonitoredTripPersistenceTest  extends OtpMiddlewareTest {
```
There's an extra space here before `extends`.

br648 replied:
Updated.

evansiroky reviewed (dismissed):
Good to go after addressing my 2 comments.

evansiroky committed:
Merge branch 'dev' into monitored-trip

evansiroky reviewed (dismissed):
I just fixed a merge conflict and pushed it, so I'm reapproving given my original request to address my 2 comments.

Commit:
style: Removed comments and white space
